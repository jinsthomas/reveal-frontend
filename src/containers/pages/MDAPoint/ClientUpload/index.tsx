import { Dictionary } from '@onaio/utils';
import { ErrorMessage, Field, Formik } from 'formik';
import React, { useState } from 'react';
import { Redirect } from 'react-router';
import { Button, Form, FormGroup, Input, Label, Modal, ModalBody, ModalHeader } from 'reactstrap';
import * as Yup from 'yup';
import LocationSelect from '../../../../components/forms/LocationSelect';
import SimpleOrgSelect from '../../../../components/forms/SimpleOrgSelect';
import LinkAsButton from '../../../../components/LinkAsButton';
import { CLIENT_UPLOAD_FORM, REQUIRED } from '../../../../configs/lang';
import { STUDENTS_LIST_URL } from '../../../../constants';
import { postUploadedFile } from '../ClientListView/helpers/serviceHooks';
import UploadStatus from '../ClientUploadStatus/';

export const uploadValidationSchema = Yup.object().shape({
  file: Yup.mixed().required(),
  jurisdictions: Yup.object().shape({
    id: Yup.string().required(REQUIRED),
    name: Yup.string(),
  }),
});
/** interface to describe props for ExportForm component */
export interface UploadFormField {
  file: string | null;
  jurisdictions: Dictionary;
  team: Dictionary;
}
export const ClientUpload = () => {
  const [selectedFile, setSelectedFile] = useState<any>(null);
  const [ifDoneHere, setIfDoneHere] = useState<boolean>(false);
  const defaultInitialValues: UploadFormField = {
    file: null,
    jurisdictions: {
      id: '',
      name: '',
    },
    team: {
      label: '',
      value: '',
    },
  };
  const closeUploadModal = {
    classNameProp: 'focus-investigation btn btn-primary float-right mt-0',
    text: 'Go Back',
    to: '/clients/students',
  };
  if (ifDoneHere) {
    return <Redirect to={STUDENTS_LIST_URL} />;
  }
  const setStateIfDone = () => {
    setIfDoneHere(true);
  };
  return (
    <div>
      <Modal isOpen={true}>
        <ModalHeader>{CLIENT_UPLOAD_FORM}</ModalHeader>
        <ModalBody>
          <Formik
            initialValues={defaultInitialValues}
            validationSchema={uploadValidationSchema}
            // tslint:disable-next-line: jsx-no-lambda
            onSubmit={async (values, { setSubmitting }) => {
              const setSubmittingStatus = () => setSubmitting(false);
              const data = new FormData();
              data.append('file', selectedFile);
              const uploadParams = `&location_id=${values.jurisdictions.id}&team_id=${values.team}`;
              await postUploadedFile(data, setStateIfDone, setSubmittingStatus, uploadParams);
            }}
          >
            {({ values, setFieldValue, handleSubmit, errors, isSubmitting }) => (
              <Form onSubmit={handleSubmit} data-enctype="multipart/form-data">
                <FormGroup className={'async-select-container'}>
                  <Label for={`jurisdictions-${1}-id`}>{'Geographical level to include'}</Label>
                  &nbsp;
                  <div style={{ display: 'inline-block', width: '24rem' }}>
                    <Field
                      required={true}
                      component={LocationSelect}
                      cascadingSelect={true}
                      name={`jurisdictions.id`}
                      id={`jurisdictions-id`}
                      className={'async-select'}
                      labelFieldName={`jurisdictions.name`}
                    />
                  </div>
                  <Field type="hidden" name={`jurisdictions.name`} id={`jurisdictions-name`} />
                  {errors.jurisdictions && (
                    <small className="form-text text-danger jurisdictions-error">
                      {'Please select location'}
                    </small>
                  )}
                  {
                    <ErrorMessage
                      name={`jurisdictions.id`}
                      component="small"
                      className="form-text text-danger"
                    />
                  }
                </FormGroup>
                {values && values.jurisdictions && values.jurisdictions.id && (
                  <FormGroup>
                    <Label for="team">Assign team for this school</Label>
                    <div style={{ display: 'inline-block', width: '24rem' }}>
                      <Field
                        required={true}
                        component={SimpleOrgSelect}
                        cascadingSelect={true}
                        name={`team`}
                        id={`team-id`}
                        className={'async-select'}
                      />
                    </div>
                  </FormGroup>
                )}
                <FormGroup>
                  <Label for="upload-file">Upload File</Label>
                  <Input
                    type="file"
                    name="file"
                    id="exampleFile"
                    accept=".csv"
                    // tslint:disable-next-line: jsx-no-lambda
                    onChange={(event: any) => {
                      setSelectedFile(event.target.files[0]);
                      setFieldValue(
                        'file',
                        event && event.target && event.target.files && event.target.files[0]
                      );
                    }}
                  />
                </FormGroup>
                <UploadStatus uploadFile={values.file} />

                {errors && errors.file ? (
                  <small className="form-text text-danger jurisdictions-error">{errors.file}</small>
                ) : null}
                <hr />
                <div style={{ display: 'inline-block', width: '12rem' }}>
                  <Button
                    type="submit"
                    id="studentexportform-submit-button"
                    className="btn btn-md btn btn-primary"
                    color="primary"
                    disabled={isSubmitting}
                  >
                    Submit
                  </Button>
                  <LinkAsButton {...closeUploadModal} />
                </div>
              </Form>
            )}
          </Formik>
        </ModalBody>
      </Modal>
    </div>
  );
};
export default ClientUpload;
