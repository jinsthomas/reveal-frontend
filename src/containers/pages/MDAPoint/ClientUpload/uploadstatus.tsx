import { Dictionary } from '@onaio/utils';
import React, { useState } from 'react';
/**
 * UploadStatus Shows status of selected file on upload
 * Todo include more specific type for props
 */
export const UploadStatus = (props: Dictionary) => {
  const [loading, setLoading] = useState<boolean>(false);
  React.useEffect(() => {
    if (!props.file) {
      return;
    }
    setLoading(true);
    const reader = new FileReader();
    /**
     * once selectedfile load setloading to false
     */
    reader.onloadend = () => {
      setLoading(false);
    };
    reader.readAsDataURL(props.file);
  }, [props.file]);

  const { file } = props;

  if (!file) {
    return null;
  }

  if (loading) {
    return <p>loading...</p>;
  }

  return <p>File is ready to submit</p>;
};
