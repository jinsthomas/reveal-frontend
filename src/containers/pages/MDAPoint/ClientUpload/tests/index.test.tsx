import { cleanup } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import ClientUpload, { defaultProps } from '..';
const history = createBrowserHistory();

describe('components/ClientUpload', () => {
  afterEach(cleanup);
  it('renders without crashing', () => {
    shallow(<ClientUpload />);
  });

  it('renders props correctly', () => {
    const props = {
      ...defaultProps,
    };
    const wrapper = mount(
      <Router history={history}>
        <ClientUpload {...props} />
      </Router>
    );
    expect(wrapper.children().props()).toMatchSnapshot('ClientUpload Select Props ownProps');
    wrapper.unmount();
  });
});
