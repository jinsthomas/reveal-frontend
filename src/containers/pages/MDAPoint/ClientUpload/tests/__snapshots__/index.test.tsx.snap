// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/LinkToNewPlans Matches snapshot 1`] = `
<ClientUpload>
  <div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      fade={true}
      isOpen={true}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      role="dialog"
      zIndex={1050}
    >
      <Portal
        node={
          <div
            style="position: relative; z-index: 1050;"
            tabindex="-1"
          >
            <div
              class=""
            >
              <div
                class="modal fade"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog"
                  role="document"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      class="modal-header"
                    >
                      <h5
                        class="modal-title"
                      >
                        Modal title
                      </h5>
                    </div>
                    <div
                      class="modal-body"
                    >
                      <form
                        data-enctype="multipart/form-data"
                      >
                        <div
                          class="form-group"
                        >
                          <label
                            class=""
                            for="upload-file"
                          >
                            Upload File
                          </label>
                          <input
                            accept=".csv"
                            class="form-control-file"
                            id="exampleFile"
                            name="file"
                            type="file"
                          />
                        </div>
                        <hr />
                        <div
                          style="display: inline-block; width: 12rem;"
                        >
                          <button
                            class="btn btn-md btn btn-primary btn btn-primary"
                            id="studentexportform-submit-button"
                            type="submit"
                          >
                            Submit
                          </button>
                          <a
                            class="focus-investigation btn btn-primary float-right mt-0"
                            href="/clients/students"
                          >
                            Go Back
                          </a>
                        </div>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-backdrop fade"
              />
            </div>
          </div>
        }
      >
        <Portal
          containerInfo={
            <div
              style="position: relative; z-index: 1050;"
              tabindex="-1"
            >
              <div
                class=""
              >
                <div
                  class="modal fade"
                  role="dialog"
                  style="display: block;"
                  tabindex="-1"
                >
                  <div
                    class="modal-dialog"
                    role="document"
                  >
                    <div
                      class="modal-content"
                    >
                      <div
                        class="modal-header"
                      >
                        <h5
                          class="modal-title"
                        >
                          Modal title
                        </h5>
                      </div>
                      <div
                        class="modal-body"
                      >
                        <form
                          data-enctype="multipart/form-data"
                        >
                          <div
                            class="form-group"
                          >
                            <label
                              class=""
                              for="upload-file"
                            >
                              Upload File
                            </label>
                            <input
                              accept=".csv"
                              class="form-control-file"
                              id="exampleFile"
                              name="file"
                              type="file"
                            />
                          </div>
                          <hr />
                          <div
                            style="display: inline-block; width: 12rem;"
                          >
                            <button
                              class="btn btn-md btn btn-primary btn btn-primary"
                              id="studentexportform-submit-button"
                              type="submit"
                            >
                              Submit
                            </button>
                            <a
                              class="focus-investigation btn btn-primary float-right mt-0"
                              href="/clients/students"
                            >
                              Go Back
                            </a>
                          </div>
                        </form>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="modal-backdrop fade"
                />
              </div>
            </div>
          }
        >
          <div
            className=""
          >
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onClick={[Function]}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              role="dialog"
              style={
                Object {
                  "display": "block",
                }
              }
              tabIndex="-1"
              tag="div"
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div
                  className="modal fade"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                  tabIndex="-1"
                >
                  <div
                    className="modal-dialog"
                    role="document"
                  >
                    <div
                      className="modal-content"
                    >
                      <ModalHeader
                        charCode={215}
                        closeAriaLabel="Close"
                        tag="h5"
                        wrapTag="div"
                      >
                        <div
                          className="modal-header"
                        >
                          <h5
                            className="modal-title"
                          >
                            Modal title
                          </h5>
                        </div>
                      </ModalHeader>
                      <ModalBody
                        tag="div"
                      >
                        <div
                          className="modal-body"
                        >
                          <Formik
                            enableReinitialize={false}
                            initialValues={
                              Object {
                                "file": null,
                                "location": undefined,
                              }
                            }
                            isInitialValid={false}
                            onSubmit={[Function]}
                            validateOnBlur={true}
                            validateOnChange={true}
                            validationSchema={
                              ObjectSchema {
                                "_blacklist": RefSet {
                                  "list": Set {},
                                  "refs": Map {},
                                },
                                "_conditions": Array [],
                                "_defaultDefault": [Function],
                                "_deps": Array [],
                                "_excludedEdges": Array [],
                                "_exclusive": Object {},
                                "_mutate": undefined,
                                "_nodes": Array [
                                  "file",
                                ],
                                "_options": Object {
                                  "abortEarly": true,
                                  "recursive": true,
                                },
                                "_type": "object",
                                "_typeError": [Function],
                                "_whitelist": RefSet {
                                  "list": Set {},
                                  "refs": Map {},
                                },
                                "fields": Object {
                                  "file": SchemaType {
                                    "_blacklist": RefSet {
                                      "list": Set {},
                                      "refs": Map {},
                                    },
                                    "_conditions": Array [],
                                    "_deps": Array [],
                                    "_exclusive": Object {
                                      "required": true,
                                    },
                                    "_mutate": undefined,
                                    "_options": Object {
                                      "abortEarly": true,
                                      "recursive": true,
                                    },
                                    "_type": "mixed",
                                    "_typeError": [Function],
                                    "_whitelist": RefSet {
                                      "list": Set {},
                                      "refs": Map {},
                                    },
                                    "tests": Array [
                                      [Function],
                                    ],
                                    "transforms": Array [],
                                  },
                                },
                                "tests": Array [],
                                "transforms": Array [
                                  [Function],
                                ],
                              }
                            }
                          >
                            <form
                              data-enctype="multipart/form-data"
                              onSubmit={[Function]}
                            >
                              <FormGroup
                                tag="div"
                              >
                                <div
                                  className="form-group"
                                >
                                  <Label
                                    for="upload-file"
                                    tag="label"
                                    widths={
                                      Array [
                                        "xs",
                                        "sm",
                                        "md",
                                        "lg",
                                        "xl",
                                      ]
                                    }
                                  >
                                    <label
                                      className=""
                                      htmlFor="upload-file"
                                    >
                                      Upload File
                                    </label>
                                  </Label>
                                  <Input
                                    accept=".csv"
                                    id="exampleFile"
                                    name="file"
                                    onChange={[Function]}
                                    type="file"
                                  >
                                    <input
                                      accept=".csv"
                                      className="form-control-file"
                                      id="exampleFile"
                                      name="file"
                                      onChange={[Function]}
                                      type="file"
                                    />
                                  </Input>
                                </div>
                              </FormGroup>
                              <UploadStatus
                                uploadFile={null}
                              />
                              <hr />
                              <div
                                style={
                                  Object {
                                    "display": "inline-block",
                                    "width": "12rem",
                                  }
                                }
                              >
                                <Button
                                  className="btn btn-md btn btn-primary"
                                  color="primary"
                                  id="studentexportform-submit-button"
                                  tag="button"
                                  type="submit"
                                >
                                  <button
                                    aria-label={null}
                                    className="btn btn-md btn btn-primary btn btn-primary"
                                    id="studentexportform-submit-button"
                                    onClick={[Function]}
                                    type="submit"
                                  >
                                    Submit
                                  </button>
                                </Button>
                                <LinkAsButton
                                  classNameProp="focus-investigation btn btn-primary float-right mt-0"
                                  text="Go Back"
                                  to="/clients/students"
                                >
                                  <Link
                                    className="focus-investigation btn btn-primary float-right mt-0"
                                    replace={false}
                                    to="/clients/students"
                                  >
                                    <a
                                      className="focus-investigation btn btn-primary float-right mt-0"
                                      href="/clients/students"
                                      onClick={[Function]}
                                    >
                                      Go Back
                                    </a>
                                  </Link>
                                </LinkAsButton>
                              </div>
                            </form>
                          </Formik>
                        </div>
                      </ModalBody>
                    </div>
                  </div>
                </div>
              </Transition>
            </Fade>
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal-backdrop"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={true}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              tag="div"
              timeout={150}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={150}
                unmountOnExit={false}
              >
                <div
                  className="modal-backdrop fade"
                />
              </Transition>
            </Fade>
          </div>
        </Portal>
      </Portal>
    </Modal>
  </div>
</ClientUpload>
`;
